prov/shm: refactor unexpected path to fix multi_recv messaging

aingerson opened the pull request:
Fixes #8166

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,34 +94,80 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
 out:
 	ofi_spin_unlock(&srx->lock);
 	return ret;
+
```
random extra blank line added

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,34 +94,80 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+			unexp->peer_entry.owner_context = owner_entry;
+		}
+
+		if (owner_entry && smr_adjust_multi_recv(srx,
```
Do you need a lock around smr_adjust_multi_recv()?

aingerson commented:
Updated to remove blank line and fix locking. Also refactored the adjust/iov setting code a bit to make it more readable.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+			unexp->peer_entry.owner_context = owner_entry;
+		}
+
+		if (owner_entry) {
```
You have back to back if statements that check the same thing.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
-	ssize_t ret = -FI_EAGAIN;
-	fi_addr_t use_addr;
+	struct smr_match_attr match_attr;
+	struct smr_rx_entry *unexp, *owner_entry = NULL;
```
This function is hard to follow.  I think adding an *mrecv_entry and using that instead of owner_entry would help.  The code has a lot of checks around owner_entry being non-null, but what it is actually trying to check is if the posted receive is a multi-recv entry.  owner_entry is used for both multi-recv and normal receives, but switches its meaning 3/4th of the way through the function after the while loop.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+			if (smr_adjust_multi_recv(srx, &owner_entry->peer_entry,
+						  unexp->peer_entry.size))
+				buf_done = true;
+			iov = owner_entry->iov;
```
Don't you need this before you adjust the mrecv buffer?

aingerson replied:
the iov is saved for the next message/for posting if we can't progress any unexpected messages. the owner_entry iov is used to initialize the unexpected rx_entry. some refactoring is obviously needed so I'll smooth this out and repush

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+		ret = srx->peer_srx.peer_ops->start_msg(&unexp->peer_entry);
+		if (ret)
+			return ret;
+		else if (!owner_entry || buf_done)
```
nit: else isn't needed as if statement above returns

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+		if (ret)
+			return ret;
+		else if (!owner_entry || buf_done)
+			return FI_SUCCESS;
```
If done_buf is true, what happened to the original owner_entry?  Was that lost?

aingerson replied:
The owner entry is freed in (free_entry) which is called in the start_msg processing/whenever the message is done so we don't care about freeing it here.

aingerson commented:
@shefty Updated to separate regular and mrecv paths. The unexpected message processing is different enough that it gets really hard to read if we keep them together. It also helps remove that complexity from the tagged path which does not use multi recv. Let me know what you think.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -131,10 +198,14 @@ static ssize_t smr_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
+	if (flags & FI_MULTI_RECV)
+		return smr_generic_mrecv(smr_get_smr_srx(ep),
+					 msg->msg_iov, msg->desc, msg->iov_count, 
+					 msg->addr, msg->context, flags);
```
Why not move the FI_MULTI_RECV check info smr_generic_recv, rather than duplicating it before each call to it?

aingerson replied:
I had it like that before but moved it to remove the check from the tagged path which never uses multi recv and previously an app could call tagged + multi recv and it would pass but have issues later. I can just add an assert though instead.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,30 +94,97 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+			return -FI_ENOMEM;
+
+		dlist_insert_tail((struct dlist_entry *) (&rx_entry->peer_entry),
+				  &recv_queue->list);
```
Need a lock here.  Maybe also around smr_get_recv_entry()?

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,30 +94,97 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+	if (ret)
+		return ret;
+
+	return FI_SUCCESS;
```
Can just return ret and skip the if statement.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -94,30 +94,97 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
+	mrecv_entry = smr_get_recv_entry(srx, iov, desc, iov_count, addr,
+					 context, 0, 0, flags);
+	mrecv_entry->peer_entry.size = ofi_total_iov_len(iov, iov_count);
+	if (!mrecv_entry) {
```
It doesn't do any good to check that mrecv_entry is valid after the call before has already dereferenced it.

aingerson commented:
Updated to move mrecv check inside generic_recv and other fixes. @shefty thanks for catching that missing lock. I've added the lock back in for that code but don't love the locking/unlocking flow. couldn't really see a better way to do it. let me know if you have a better idea. (I'm talking about lines 178-190)

shefty reviewed (approved):
Looks fine.  Locking is a little odd looking, but easy to follow, and I don't see an obvious way to improve it.
